montecarlo/eg/test/pdg.cxx

```diff
@@ -0,0 +1,80 @@
+// Copyright (C) 1995-2023, Rene Brun and Fons Rademakers.
+// All rights reserved.
+//
+// For the licensing terms see $ROOTSYS/LICENSE.
+// For the list of contributors see $ROOTSYS/README/CREDITS.
+
+// @author Danilo Piparo CERN
+
+#include <TDatabasePDG.h>
+#ifdef R__USE_IMT
+#include <ROOT/TThreadExecutor.hxx>
+#endif
+
+#include "gtest/gtest.h"
+
+#include <vector>
+
+#ifndef R__B64
+static const auto NUM_THREADS = 8u;
+#else
+static const auto NUM_THREADS = 0u;
+#endif
+
+#ifdef R__USE_IMT
+
+class TPDGTable : public ::testing::Test {
+protected:
+   TPDGTable() : fParticleCodes(fParticlesNames.size()), fInstance(TDatabasePDG::Instance()), fPool(NUM_THREADS)
+   {
+
+      std::transform(fParticlesNames.begin(), fParticlesNames.end(), fParticleCodes.begin(),
+                     [this](const char *pname) { return this->fInstance->GetParticle(pname)->PdgCode(); });
+   }
+   static constexpr std::initializer_list<const char *> fParticlesNames = {
+      // Mesons
+      "pi0", "rho0", "a_20", "K_L0", "pi_diffr+", "pi+", "rho+", "a_2+", "omega_di", "eta", "omega", "f_2", "K_S0",
+      "K0", "K*0", "K*_20", "K+", "K*+", "K*_2+", "phi_diff", "eta'", "phi", "f'_2", "D+", "D*+", "D*_2+", "D0", "D*0",
+      "D*_20", "D_s+", "D*_s+", "D*_2s+", "J/psi_di", "eta_c", "J/psi", "chi_2c", "B0", "B*0", "B*_20", "B+", "B*+",
+      "B*_2+", "B_s0", "B*_s0", "B*_2s0", "B_c+", "B*_c+", "B*_2c+", "eta_b", "Upsilon", "chi_2b",
+      // Baryons
+      "neutron", "Delta0", "proton", "Delta+", "Delta++", "Sigma-", "Sigma*-", "Lambda0", "Sigma0", "Sigma*0", "Sigma+",
+      "Sigma*+", "Xi-", "Xi*-", "Xi0", "Xi*0", "Omega-", "Sigma_c0", "Sigma*_c0", "Lambda_c+", "Xi_c0", "Sigma_c+",
+      "Sigma*_c+", "Sigma_c++", "Sigma*_c++", "Xi_c+", "cs_0", "Xi'_c0", "Xi*_c0", "Xi'_c+", "Xi*_c+", "Omega_c0",
+      "Omega*_c0", "Xi_cc+", "Xi*_cc+", "Xi_cc++", "Xi*_cc++", "Omega_cc+", "Omega*_cc+", "Omega*_ccc++", "Sigma_b-",
+      "Sigma*_b-", "Lambda_b0", "Xi_b-", "Xi_bc0", "Sigma_b0", "Sigma*_b0", "Sigma_b+", "Sigma*_b+", "Xi_b0", "Xi_bc+",
+      "Xi'_b-", "Xi*_b-", "Xi'_b0", "Xi*_b0", "Omega_b-", "Omega*_b-", "Omega_bc0", "Xi'_bc0", "Xi*_bc0", "Xi'_bc+",
+      "Xi*_bc+", "Omega'_bc0", "Omega*_bc0", "Omega_bcc+", "Omega*_bcc+", "Xi*_bb-", "Xi_bb0", "Xi*_bb0", "Omega_bb-",
+      "Omega*_bb-", "Omega_bbc0", "Omega*_bbc0", "Omega*_bbb-", "Lambda(1520)0",
+      // Bosons
+      "g", "gamma", "Z0", "W+", "h0", "reggeon", "pomeron", "Z'0", "Z\"0", "W'+", "H0", "A0", "H+",
+      // Leptons
+      "e-", "nu_e", "mu-", "nu_mu", "tau-", "nu_tau", "tau'-", "nu'_tau"};
+
+   std::vector<int> fParticleCodes;
+   TDatabasePDG *fInstance;
+   ROOT::TThreadExecutor fPool;
+};
+
+TEST_F(TPDGTable, ConcurrentGetParticleByNameCalls)
```

Comment:
I am not clear on the semantic of `TEST_F` here but wouldn't this means that the `TPDGTable` object will be created first and isn't the constructor of `TPDGTable` already (indirectly) initializing the pdg table?  I.e. Do we test what we mean to test here?

Reply:
hi @pcanal. You are absolutely right. I think I understood from the report that the 2 GetParticle methods needed to be thread safe and this is what this PR enables and tests. Making the static function `TDatabasePDG::Instance()` also ts is of course possible, even if I think was not requested (and somewhat feels like swimming against the current, given that having TS getters is perhaps natural, while initialisation less so, in general). Of course I am open to any suggestion, given that technically the cost to implement more features is quite low.

Reply:
>? Making the static function TDatabasePDG::Instance() also ts is of course possible

It of course should be but indeed should not be really part of this PR :).

What I observe is that the test is currently not testing the thread safety of `GetParticle`.

`TPDGTable` is created first (before threads) and its constructor contains:
```
this->fInstance->GetParticle(pname)->PdgCode();
```
consequently, (If I read correctly), `TDatabasePDG::fParticleList` will be set then and subsequent calls to `GetParticle` will work properly from multiple threads with or without this PR.

Am I missing something?